src/views/virtualmachines/tree/utils/utils.tsx

```diff
@@ -219,18 +270,120 @@ export const createTreeViewData = (
-    ? createAllNamespacesTreeItem(getResourceUrl, treeViewData, treeViewDataMap, projectMap)
-    : null;
+  if (clusterName) {
+    return [
+      createClusterLevelTreeItem(
+        getResourceUrl,
+        treeViewData,
+        treeViewDataMap,
+        projectMap,
+        clusterName,
+      ),
+    ];
+  } else {
+    const allNamespacesTreeItem = isAdmin
+      ? createAllNodesTreeItem(getResourceUrl, treeViewData, treeViewDataMap, projectMap)
+      : null;
+
+    treeDataMap.value = treeViewDataMap;
+
+    const tree = allNamespacesTreeItem ? [allNamespacesTreeItem] : treeViewData;
+
+    return tree;
+  }
+};
 
-  treeDataMap.value = treeViewDataMap;
+export const createTreeViewData = (
+  getResourceUrlMultiClusterOverride: (cluster?: string) => GetResourceUrl,
+  isAdmin: boolean,
+  pathname: string,
+  foldersEnabled: boolean,
+  projectNames?: MulticlusterResource<K8sResourceCommon>[] | string[],
+  vms?: MulticlusterResource<V1VirtualMachine>[] | V1VirtualMachine[],
+  supportsMulticluster?: boolean,
+): TreeViewDataItem[] => {
+  const { currentVMTab, vmName } = getVMInfoFromPathname(pathname);
+  const clusterTreeItems = [];
+  const treeViewDataMap: Record<string, TreeViewDataItem> = {};
 
-  const tree = allNamespacesTreeItem ? [allNamespacesTreeItem] : treeViewData;
+  if (supportsMulticluster) {
+    const multiclusterVmsRecord: Record<string, MulticlusterResource<V1VirtualMachine>[]> = {};
+    const multiclusterVms = vms as MulticlusterResource<V1VirtualMachine>[];
 
-  return tree;
+    multiclusterVms.forEach((vm) => {
+      if (multiclusterVmsRecord[vm.cluster]) {
+        multiclusterVmsRecord[vm.cluster].push(vm);
+      } else {
+        multiclusterVmsRecord[vm.cluster] = [vm];
+      }
+    });
+
+    const namespaceNameByClusterRecord: Record<string, string[]> = {};
+    projectNames.forEach((resource) => {
+      namespaceNameByClusterRecord[resource.cluster]
+        ? namespaceNameByClusterRecord[resource.cluster].push(getName(resource))
+        : (namespaceNameByClusterRecord[resource.cluster] = [getName(resource)]);
+    });
+    const getResourceUrlDefault = getResourceUrlMultiClusterOverride();
+    Object.keys(multiclusterVmsRecord).forEach((clusterName) => {
+      const namespaceNames = namespaceNameByClusterRecord[clusterName];
+      const getResourceUrlForSpokeCluster = getResourceUrlMultiClusterOverride(clusterName);
+      const clusterVms = multiclusterVmsRecord[clusterName];
+
+      clusterTreeItems.push(
+        ...createClusterTreeViewData(
+          getResourceUrlForSpokeCluster,
+          namespaceNames,
+          clusterVms,
+          isAdmin,
+          pathname,
+          foldersEnabled,
+          treeViewDataMap,
+          clusterName,
+        ),
+      );
+    });
+    let aggrigateProjectMap = {};
+    Object.keys(multiclusterVmsRecord).forEach((clusterName) => {
+      aggrigateProjectMap = {
+        ...aggrigateProjectMap,
+        ...buildProjectMap(
```

Comment:
Move this out of the loop and try to run it once at the beginning of the function.